dlib/type_traits.h

```diff
@@ -210,8 +210,136 @@ namespace dlib
+
+// ----------------------------------------------------------------------------------------
+
+    template<class F>
+    struct callable_traits;
+    /*!
+        WHAT THIS OBJECT REPRESENTS
+            This is a type trait for callable types like function pointers, functors and lambdas.
+            It provides the following types:
+                return_type : the return type of the callable object
+                args        : a parameter pack packaged in a types_<> meta container containing
+                              all the function argument types
+            It also provides the following static members:
+                nargs       : the number of function arguments
+            
+            For example, a function type F with signature R(T1, T2, T3) has the following traits:
+                callable_traits<F>::return_type == R
+                callable_traits<F>::args        == types_<T1,T2,T3>
+                callable_traits<F>::nargs       == 3
+    !*/
+
+    template<class R, class... Args>
+    struct callable_traits<R(Args...)>
+    {
+        using return_type = R;
+        using args        = types_<Args...>;
+        constexpr static std::size_t nargs = sizeof...(Args);
+    };  
+
+    template<class R, class... Args>
+    struct callable_traits<R(*)(Args...)> : public callable_traits<R(Args...)>{};
+
+    template<class C, class R, class... Args>
+    struct callable_traits<R(C::*)(Args...)> : public callable_traits<R(Args...)>{};
+
+    template<class C, class R, class... Args>
+    struct callable_traits<R(C::*)(Args...) const> : public callable_traits<R(Args...)>{};
+
+    template<class F>
+    struct callable_traits
+    {
+        using call_type     = callable_traits<decltype(&std::decay_t<F>::operator())>;
+        using return_type   = typename call_type::return_type;
+        using args          = typename call_type::args;
+        constexpr static std::size_t nargs = call_type::nargs;
+    };
+
+    template<class Callable>
+    using callable_args = typename callable_traits<Callable>::args;
+    /*!
+        WHAT THIS OBJECT REPRESENTS
+            This is a type alias for callable_traits::args
+    !*/
+
+
+    template<std::size_t I, class Callable>
+    using callable_arg = nth_type_t<I, callable_args<Callable>>;
+    /*!
+        WHAT THIS OBJECT REPRESENTS
+            This is a type alias for nth_type_t<callable_traits::args>
+    !*/
+
+    template<class Callable>
+    using callable_nargs = std::integral_constant<std::size_t, callable_traits<Callable>::nargs>;
+    /*!
+        WHAT THIS OBJECT REPRESENTS
+            This returns a compile-time constant for callable_traits::nargs 
+            packaged in a std::integral_constant.
+    !*/
+
+    template<class Callable>
+    using callable_return = typename callable_traits<Callable>::return_type;
+    /*!
+        WHAT THIS OBJECT REPRESENTS
+            This is a type alias for callable_traits::return_type
+    !*/
```

Comment:
```suggestion
    template<class Callable>
    using callable_args = typename callable_traits<Callable>::args;

    template<std::size_t I, class Callable>
    using callable_arg = nth_type_t<I, callable_args<Callable>>;
    
    template<class Callable>
    using callable_nargs = std::integral_constant<std::size_t, callable_traits<Callable>::nargs>;

    template<class Callable>
    using callable_return = typename callable_traits<Callable>::return_type;
```
No docs when the only thing to say is to restate what the code trivially says :)